scenedetect/detectors/hash_detector.py

```diff
@@ -0,0 +1,168 @@
+
+class HashDetector(SceneDetector):
+    """Detects cuts using a perceptual hashing algorithm. For more information
+    on the perceptual hashing algorithm see references below.
+
+    1. https://www.hackerfactor.com/blog/index.php?/archives/432-Looks-Like-It.html
+    2. https://github.com/JohannesBuchner/imagehash
+
+    Since the difference between frames is used, unlike the ThresholdDetector,
+    only fast cuts are detected with this method.
+    """
+
+    def __init__(self, threshold=100.0, min_scene_len=15, hash_size=16, highfreq_factor=2):
+        super(HashDetector, self).__init__()
+        self.threshold = threshold
+        # Minimum length of any given scene, in frames (int) or FrameTimecode
+        self.min_scene_len = min_scene_len
+        # Size of square of low frequency data to include from the discrete cosine transform
+        self.hash_size = hash_size
+        # How much high frequency data should be thrown out from the DCT
+        # A value of 2 means only keep 1/2 of the freq data, a value of 4 means only keep 1/4
+        self.highfreq_factor = highfreq_factor
+        self.last_frame = None
+        self.last_scene_cut = None
+        self.last_hash = numpy.array([])
+        self._metric_keys = ['hash_dist']
+        self.cli_name = 'detect-hash'
+    
+    def get_metrics(self):
+        return self._metric_keys
+    
+    def process_frame(self, frame_num, frame_img):
+        """ Similar to ContentDetector, but using a perceptual hashing algorithm
+        to calculate a hash for each frame and then calculate a hash difference
+        frame to frame.
+
+        Arguments:
+            frame_num (int): Frame number of frame that is being passed.
+
+            frame_img (Optional[int]): Decoded frame image (numpy.ndarray) to perform scene
+                detection on. Can be None *only* if the self.is_processing_required() method
+                (inhereted from the base SceneDetector class) returns True.
+
+        Returns:
+            List[int]: List of frames where scene cuts have been detected. There may be 0
+            or more frames in the list, and not necessarily the same as frame_num.
+        """
+
+        cut_list = []
+        metric_keys = self._metric_keys
+        _unused = ''
+
+        # Initialize last scene cut point at the beginning of the frames of interest.
+        if self.last_scene_cut is None:
+            self.last_scene_cut = frame_num
+        
+        # We can only start detecting once we have a frame to compare with.
+        if self.last_frame is not None:
+            # We obtain the change in hash value between subsequent frames as 
+            # well as the actual hash value. This is refered to in a statsfile
+            # as their respective metric keys.
+            if (self.stats_manager is not None and
+                    self.stats_manager.metrics_exist(frame_num, metric_keys)):
+                hash_dist = self.stats_manager.get_metrics(frame_num, metric_keys)
```

Comment:
I think we should remove this optimization since long term plan is to get rid of loading stats in lieu of #235.  This is a good example of why we should remove it actually - since the metric key doesn't contain all of the parameters, if a statsfile is specified, changing them won't make the output any different.

Reply:
Done with 0f54efc.